File: app/src/main/java/app/bristlecone/RegionManager.java
package app.bristlecone;

import android.content.Context;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Downloads a map region (tiles, terrain, trail data, land boundaries) so it works with no
 * signal. The web layer decides which URLs make up a region; this class fetches them in the
 * background, pins them against cache trimming, and reports progress.
 */
public final class RegionManager {
    public interface Listener { void onProgress(JSONObject status); }

    private final File dir;
    private final NetCache net;
    private final Listener listener;
    private final ExecutorService pool = Executors.newFixedThreadPool(4);
    private final Map<String, AtomicBoolean> cancels = new ConcurrentHashMap<>();

    public RegionManager(Context ctx, NetCache net, Listener listener) {
        this.dir = new File(ctx.getFilesDir(), "regions");
        //noinspection ResultOfMethodCallIgnored
        dir.mkdirs();
        this.net = net;
        this.listener = listener;
        refreshPins();
    }

    public void start(final JSONObject spec) throws Exception {
        final String id = spec.getString("id");
        final JSONArray urls = spec.getJSONArray("urls");
        final int total = urls.length();
        final AtomicBoolean cancel = new AtomicBoolean(false);
        cancels.put(id, cancel);

        final JSONObject manifest = new JSONObject();
        manifest.put("id", id);
        manifest.put("name", spec.optString("name", "Region"));
        manifest.put("bbox", spec.optJSONArray("bbox"));
        manifest.put("minZoom", spec.optInt("minZoom"));
        manifest.put("maxZoom", spec.optInt("maxZoom"));
        manifest.put("layers", spec.optJSONArray("layers"));
        manifest.put("created", System.currentTimeMillis());
        manifest.put("state", "downloading");
        manifest.put("total", total);
        manifest.put("done", 0);
        manifest.put("failed", 0);
        manifest.put("bytes", 0);
        final JSONArray keys = new JSONArray();
        for (int i = 0; i < total; i++) keys.put(CachePolicy.key(urls.getString(i)));
        manifest.put("keys", keys);
        save(manifest);
        refreshPins();

        final AtomicInteger done = new AtomicInteger();
        final AtomicInteger failed = new AtomicInteger();
        final AtomicLong bytes = new AtomicLong();
        final AtomicLong lastReport = new AtomicLong();

        new Thread(new Runnable() {
            @Override public void run() {
                final Object lock = new Object();
                for (int i = 0; i < total; i++) {
                    if (cancel.get()) break;
                    final String url = urls.optString(i);
                    pool.execute(new Runnable() {
                        @Override public void run() {
                            if (!cancel.get()) {
                                try {
                                    NetCache.Result r = net.isCachedFresh(url) ? null : net.get(url, true);
                                    if (r != null) {
                                        if (r.status == 200 || r.status == 204 || r.status == 404) bytes.addAndGet(r.body.length);
                                        else failed.incrementAndGet();
                                        // Be gentle with volunteer-run data services.
                                        if (url.contains("overpass")) Thread.sleep(1200);
                                    }
                                } catch (Exception e) {
                                    failed.incrementAndGet();
                                }
                            }
                            int d = done.incrementAndGet();
                            long now = System.currentTimeMillis();
                            if (d == total || now - lastReport.get() > 400) {
                                lastReport.set(now);
                                report(manifest, d, failed.get(), bytes.get(), "downloading");
                            }
                            synchronized (lock) { lock.notifyAll(); }
                        }
                    });
                    // Keep the queue short so cancel is quick and memory stays low.
                    synchronized (lock) {
                        while (i - done.get() > 24 && !cancel.get()) {
                            try { lock.wait(250); } catch (InterruptedException ignored) { }
                        }
                    }
                }
                while (done.get() < total && !cancel.get()) {
                    try { Thread.sleep(200); } catch (InterruptedException ignored) { }
                }
                String state = cancel.get() ? "cancelled" : (failed.get() > 0 ? "partial" : "ready");
                report(manifest, done.get(), failed.get(), bytes.get(), state);
                cancels.remove(id);
                if (cancel.get()) delete(id);
            }
        }, "region-" + id).start();
    }

    private void report(JSONObject manifest, int done, int failed, long bytes, String state) {
        try {
            synchronized (manifest) {
                manifest.put("done", done);
                manifest.put("failed", failed);
                manifest.put("bytes", bytes);
                manifest.put("state", state);
                if (!"downloading".equals(state)) save(manifest);
            }
            JSONObject s = new JSONObject();
            s.put("id", manifest.getString("id"));
            s.put("name", manifest.optString("name"));
            s.put("done", done);
            s.put("total", manifest.optInt("total"));
            s.put("failed", failed);
            s.put("bytes", bytes);
            s.put("state", state);
            listener.onProgress(s);
        } catch (Exception ignored) {
        }
    }

    public void cancel(String id) {
        AtomicBoolean c = cancels.get(id);
        if (c != null) c.set(true);
    }

    public void delete(String id) {
        //noinspection ResultOfMethodCallIgnored
        new File(dir, safe(id) + ".json").delete();
        refreshPins();
        net.trim();
    }

    public JSONArray list() {
        JSONArray out = new JSONArray();
        File[] fs = dir.listFiles();
        if (fs == null) return out;
        for (File f : fs) {
            JSONObject m = load(f);
            if (m == null) continue;
            m.remove("keys");
            try {
                if (cancels.containsKey(m.optString("id"))) m.put("state", "downloading");
                else if ("downloading".equals(m.optString("state"))) m.put("state", "partial");
            } catch (Exception ignored) { }
            out.put(m);
        }
        return out;
    }

    public void refreshPins() {
        Set<String> pins = new HashSet<>();
        File[] fs = dir.listFiles();
        if (fs != null) for (File f : fs) {
            JSONObject m = load(f);
            if (m == null) continue;
            JSONArray k = m.optJSONArray("keys");
            if (k != null) for (int i = 0; i < k.length(); i++) pins.add(k.optString(i));
        }
        net.setPinned(pins);
    }

    public File dir() { return dir; }

    private void save(JSONObject m) {
        try {
            FileOutputStream o = new FileOutputStream(new File(dir, safe(m.getString("id")) + ".json"));
            o.write(m.toString().getBytes("UTF-8"));
            o.close();
        } catch (Exception ignored) {
        }
    }

    private static JSONObject load(File f) {
        if (!f.getName().endsWith(".json")) return null;
        try {
            return new JSONObject(new String(NetCache.readAll(new FileInputStream(f)), "UTF-8"));
        } catch (Exception e) {
            return null;
        }
    }

    static String safe(String id) { return id.replaceAll("[^A-Za-z0-9_-]", "_"); }
}

File: ios/Bristlecone/Sources/RegionManager.swift
import Foundation

/// Downloads a map region (tiles, terrain, trail data, land boundaries) so it works with no
/// signal. The page decides which URLs make up a region; this fetches them, pins them against
/// cache trimming and reports progress.
final class RegionManager {
    let dir: URL
    private let net: NetCache
    private let onProgress: ([String: Any]) -> Void
    private let queue = DispatchQueue(label: "bristlecone.regions")
    private var cancelled = Set<String>()
    private var active = Set<String>()

    init(root: URL, net: NetCache, onProgress: @escaping ([String: Any]) -> Void) {
        dir = root.appendingPathComponent("regions", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        self.net = net
        self.onProgress = onProgress
        refreshPins()
    }

    static func safe(_ id: String) -> String {
        String(id.map { c -> Character in c.isASCII && (c.isLetter || c.isNumber || c == "_" || c == "-") ? c : "_" })
    }

    private func file(_ id: String) -> URL { dir.appendingPathComponent(RegionManager.safe(id) + ".json") }

    func start(_ spec: [String: Any]) -> String {
        guard let id = spec["id"] as? String, let urls = spec["urls"] as? [String] else { return "error: bad region" }
        let manifest: [String: Any] = [
            "id": id, "name": spec["name"] as? String ?? "Region",
            "bbox": spec["bbox"] ?? NSNull(), "minZoom": spec["minZoom"] ?? 0, "maxZoom": spec["maxZoom"] ?? 0,
            "layers": spec["layers"] ?? [Any](), "created": Int64(NetCache.now()), "state": "downloading",
            "total": urls.count, "done": 0, "failed": 0, "bytes": 0,
            "keys": urls.map(CachePolicy.key)
        ]
        save(manifest)
        refreshPins()
        queue.sync { _ = active.insert(id); _ = cancelled.remove(id) }

        DispatchQueue.global(qos: .utility).async { [weak self] in
            guard let self = self else { return }
            let slots = DispatchSemaphore(value: 4)
            let group = DispatchGroup()
            let counters = NSLock()
            var done = 0, failed = 0
            var bytes: Int64 = 0
            var lastReport = 0.0
            for url in urls {
                if self.isCancelled(id) { break }
                slots.wait()
                group.enter()
                let finishOne = { (ok: Bool, size: Int) in
                    counters.lock()
                    done += 1
                    if ok { bytes += Int64(size) } else { failed += 1 }
                    let d = done, f = failed, b = bytes
                    let now = NetCache.now()
                    let report = d == urls.count || now - lastReport > 400
                    if report { lastReport = now }
                    counters.unlock()
                    if report { self.report(manifest, done: d, failed: f, bytes: b, state: "downloading") }
                    slots.signal()
                    group.leave()
                }
                if self.net.isCachedFresh(url) { finishOne(true, 0); continue }
                self.net.get(url, forceNetwork: true) { r in
                    let ok = r.status == 200 || r.status == 204 || r.status == 404
                    // Be gentle with volunteer-run data services.
                    let delay: Double = url.contains("overpass") ? 1.2 : 0
                    DispatchQueue.global(qos: .utility).asyncAfter(deadline: .now() + delay) { finishOne(ok, r.body.count) }
                }
            }
            group.wait()
            let wasCancelled = self.isCancelled(id)
            counters.lock()
            let d = done, f = failed, b = bytes
            counters.unlock()
            self.report(manifest, done: d, failed: f, bytes: b, state: wasCancelled ? "cancelled" : (f > 0 ? "partial" : "ready"), finished: true)
            self.queue.sync { _ = self.active.remove(id) }
            if wasCancelled { self.delete(id) }
        }
        return "ok"
    }

    private func isCancelled(_ id: String) -> Bool { queue.sync { cancelled.contains(id) } }

    private func report(_ base: [String: Any], done: Int, failed: Int, bytes: Int64, state: String, finished: Bool = false) {
        var manifest = base
        manifest["done"] = done; manifest["failed"] = failed; manifest["bytes"] = bytes; manifest["state"] = state
        if finished { save(manifest) }
        onProgress(["id": manifest["id"] ?? "", "name": manifest["name"] ?? "", "done": done,
                    "total": manifest["total"] ?? 0, "failed": failed, "bytes": bytes, "state": state])
    }

    func cancel(_ id: String) { queue.sync { _ = cancelled.insert(id) } }

    func delete(_ id: String) {
        try? FileManager.default.removeItem(at: file(id))
        refreshPins()
        net.trim()
    }

    func list() -> [[String: Any]] {
        let names = (try? FileManager.default.contentsOfDirectory(atPath: dir.path)) ?? []
        let running = queue.sync { active }
        return names.compactMap { n -> [String: Any]? in
            guard var m = load(dir.appendingPathComponent(n)) else { return nil }
            m.removeValue(forKey: "keys")
            if let id = m["id"] as? String, running.contains(id) { m["state"] = "downloading" }
            else if m["state"] as? String == "downloading" { m["state"] = "partial" }
            return m
        }
    }

    func refreshPins() {
        var pins = Set<String>()
        for n in (try? FileManager.default.contentsOfDirectory(atPath: dir.path)) ?? [] {
            if let keys = load(dir.appendingPathComponent(n))?["keys"] as? [String] { pins.formUnion(keys) }
        }
        net.pinned = pins
    }

    private func save(_ m: [String: Any]) {
        guard let id = m["id"] as? String, let d = try? JSONSerialization.data(withJSONObject: m) else { return }
        try? d.write(to: file(id), options: .atomic)
    }

    private func load(_ f: URL) -> [String: Any]? {
        guard f.pathExtension == "json", let d = try? Data(contentsOf: f) else { return nil }
        return (try? JSONSerialization.jsonObject(with: d)) as? [String: Any]
    }
}
